desktop-widgets/divesiteimportdialog.h

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: GPL-2.0
+#ifndef DIVESITEIMPORTDIALOG_H
+#define DIVESITEIMPORTDIALOG_H
+
+#include <QDialog>
+#include <QThread>
+#include <QHash>
+#include <QMap>
+#include <QAbstractTableModel>
+#include <memory>
+
+#include "ui_divesiteimportdialog.h"
+#include "core/divesite.h"
+
+namespace Ui {
+	class DivesiteImportDialog;
+}
+
+class DivesiteImportedModel;
+
+class DivesiteImportDialog : public QDialog {
+	Q_OBJECT
+public:
+	explicit DivesiteImportDialog(QWidget *parent = 0, struct dive_site_table *imported = NULL, QString source = "");
```

Comment:
@bstoeger I want to be sure I understand what you are saying in your second comment above.

I believe what you are saying is that since `DivesiteImportDialog` is being passed a pointer to a `struct dive_site_table`, it should not be the responsibility of `DivesiteImportDialog` to free the memory used by the table -- the code that called (or created) the `DivesiteImportDialog` object should be responsible for freeing the memory.

Regarding the second part of your comment about the table being cleared in the destructor, I'm not sure what you are asking.

Reply:
>  believe what you are saying is that since DivesiteImportDialog is being passed a pointer to a struct dive_site_table, it should not be the responsibility of DivesiteImportDialog to free the memory used by the table -- the code that called (or created) the DivesiteImportDialog object should be responsible for freeing the memory.

Actually I'm not saying what it should do, I'm just asking to comment what it does. E.g.
```
// Caller keeps ownership of "imported". The contents of "imported" will be consumed on execution of the dialog.
// On return, it will be empty.
```

The problem is that ownership semantics are very subtle here. What I think it *should* do, but that is not a requirement for this commit:
The class should use its own table as a subobject:
```
class DivesiteImportDialog {
private:
	struct dive_site_table imported_sites;
	...
};
```
and take ownership of the contents in the constructor:
```
// Consumes all elements of "imported". Table will be empty on return of the constructor
DiveSiteImportDialog::DiveSiteImportDialog(QWidget *parent = 0, struct dive_site_table *imported, QString source)
{
	importedSites = imported;
	imported.nr = imported.allocated = 0;
	imported.dive_sites = nullptr;
}
```

And then in the destructor clean up:
```
DiveSiteImportDialog::~DiveSiteImportDialog()
{
	clear_dive_site_table(&importedSites);
}
```
That's a poor man's implementation of "move semantics".
But that would of course mean changing all `importedSites->` to `importedSites.`, etc. So really not needed for now.

Just another comment: By convention, the `QObject *parent` should always be the last argument.

Reply:
OK. I think I understand how you are saying you would do it, but I want to make sure of one thing. In your third code block above, in the definition of the constructor, I believe it should be `struct dive_site_table imported` instead of `struct dive_site_table *imported` since you are then assigning it to the `struct dive_site_table` subobject you declared in the header file in the second code block.

I also wanted to verify in your examples that `imported_sites` in the second code block should be the same name as `importedSites` in the last two code blocks. Is `imported_sites` the preferred name since it is a C structure instead of a C++ object?

Reply:
> In your third code block above, in the definition of the constructor, I believe it should be `struct dive_site_table imported` instead of `struct dive_site_table *imported` since you are then assigning it to the `struct dive_site_table` subobject you declared in the header file in the second code block.

Yes, two possibilities:
```
DiveSiteImportDialog::DiveSiteImportDialog(QWidget *parent = 0, struct dive_site_table *imported, QString source)
{
	importedSites = *imported;
```
or
```
DiveSiteImportDialog::DiveSiteImportDialog(QWidget *parent = 0, struct dive_site_table &imported, QString source)
{
	importedSites = imported;
```

`struct dive_site_table imported` as argument type would be wrong, because it being a "dumb" C-object, the compiler would just copy the innards and now you have two tables using the same allocated memory -> crash by double-free().

But I wouldn't mind if you keep it as it is for now.

> Is imported_sites the preferred name since it is a C structure instead of a C++ object?

I think generally we use camelCase when in C++ code. But we're not very consistent about it. Better ask @dirkhh. :)